File: src/main/java/edu/cornell/cs/apl/viaduct/parsing/Parsing.java
package edu.cornell.cs.apl.viaduct.parsing;

import edu.cornell.cs.apl.viaduct.syntax.surface.ProgramNode;
import java_cup.runtime.ComplexSymbolFactory;

class Parsing {
  private Parsing() {}

  /** Parses the source file and returns the AST. */
  static ProgramNode parse(SourceFile file) throws Exception {
    ComplexSymbolFactory symbolFactory = new ComplexSymbolFactory();
    Lexer scanner = new Lexer(file, symbolFactory);
    Parser parser = new Parser(scanner, symbolFactory);
    return (ProgramNode) parser.parse().value;
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/parsing/Parsing.kt
package edu.cornell.cs.apl.viaduct.parsing

import edu.cornell.cs.apl.viaduct.security.Label
import edu.cornell.cs.apl.viaduct.syntax.surface.ProgramNode
import java_cup.runtime.ComplexSymbolFactory

/** Parses [this] string and returns the AST. */
fun String.parse(path: String = "<string>"): ProgramNode {
    return SourceFile.from(path, this).parse()
}

/** Parses [this] source file and returns the AST. */
fun SourceFile.parse(): ProgramNode {
    val symbolFactory = ComplexSymbolFactory()
    val scanner = Lexer(this, symbolFactory)
    val parser = Parser(scanner, symbolFactory)
    return parser.parseProgram()
}

/** Parses [this] string as a security label. */
fun String.parseLabel(path: String = "<string>"): Label {
    return SourceFile.from(path, this).parseLabel()
}

/** Parses [this] source file as a security label. */
fun SourceFile.parseLabel(): Label {
    val symbolFactory = ComplexSymbolFactory()
    val scanner = Lexer(this, symbolFactory)
    val parser = Parser(scanner, symbolFactory)
    return parser.parseLabel()
}
